core/src/main/java/de/danoeh/antennapod/core/sync/EpisodeActionFilter.java

```diff
@@ -0,0 +1,88 @@
+        for (EpisodeAction remoteAction : remoteActions) {
+            Log.d(TAG, "Processing remoteAction: " + remoteAction.toString());
+            Pair<String, String> key = new Pair<>(remoteAction.getPodcast(), remoteAction.getEpisode());
+            switch (remoteAction.getAction()) {
+                case NEW:
+                    remoteActionsThatOverrideLocalActions.put(key, remoteAction);
+                    break;
+                case DOWNLOAD:
+                    break;
+                case PLAY:
+                    EpisodeAction localMostRecent = localMostRecentPlayActions.get(key);
+                    if (localActionHappenedAfterRemoteAction(remoteAction, localMostRecent)) {
+                        break;
+                    }
+                    EpisodeAction mostRecentAction = remoteActionsThatOverrideLocalActions.get(key);
+                    if (mostRecentAction == null
+                            || mostRecentAction.getTimestamp() == null
+                            || remoteActionHappendAfterLocalAction(remoteAction, mostRecentAction)
+                    ) {
+                        remoteActionsThatOverrideLocalActions.put(key, remoteAction);
+                    }
+                    break;
+                case DELETE:
+                    // NEVER EVER call DBWriter.deleteFeedMediaOfItem() here, leads to an infinite loop
+                    break;
+                default:
+                    Log.e(TAG, "Unknown remoteAction: " + remoteAction);
+                    break;
+            }
+        }
+
+        return remoteActionsThatOverrideLocalActions;
+    }
+
+    private static boolean remoteActionHappendAfterLocalAction(EpisodeAction remoteAction,
+                                                               EpisodeAction mostRecentAction) {
+        return remoteAction.getTimestamp() != null
+                && mostRecentAction.getTimestamp().before(remoteAction.getTimestamp());
+    }
+
+    private static Map<Pair<String, String>, EpisodeAction> createUniqueLocalMostRecentPlayActions(
+            List<EpisodeAction> queuedEpisodeActions
+    ) {
+        Map<Pair<String, String>, EpisodeAction> localMostRecentPlayAction;
+        localMostRecentPlayAction = new ArrayMap<>();
+        for (EpisodeAction action : queuedEpisodeActions) {
+            Pair<String, String> key = new Pair<>(action.getPodcast(), action.getEpisode());
+            EpisodeAction mostRecent = localMostRecentPlayAction.get(key);
+            if (mostRecent == null || mostRecent.getTimestamp() == null) {
+                localMostRecentPlayAction.put(key, action);
+            } else if (mostRecent.getTimestamp().before(action.getTimestamp())) {
+                localMostRecentPlayAction.put(key, action);
+            }
+        }
+        return localMostRecentPlayAction;
+    }
+
+    private static boolean localActionHappenedAfterRemoteAction(EpisodeAction remoteAction,
```

Comment:
Couldn't you have a single "happenedAfter" method instead of the two methods that basically just are the inverse of each other?

Reply:
I need to write tests for that to see if they really are only inverse or if there is more to it. The null checks might be an issue

Reply:
I added tests and then refactored the code yet again. I think now its clearer (at least for me) what this code does. The inverted condition is gone now too.
What do you think?